File: lib/src/screens/permissions/permissions_screen.dart
import 'package:flutter/material.dart';
import '../../services/permission_service.dart';

/// Screen shown after login to request necessary permissions
class PermissionsScreen extends StatefulWidget {
  const PermissionsScreen({super.key});

  static const routeName = '/permissions';

  @override
  State<PermissionsScreen> createState() => _PermissionsScreenState();
}

class _PermissionsScreenState extends State<PermissionsScreen> {
  final PermissionService _permissionService = PermissionService();
  bool _isRequesting = false;
  List<String> _deniedPermissions = [];

  @override
  void initState() {
    super.initState();
    _checkPermissions();
  }

  Future<void> _checkPermissions() async {
    final hasAll = await _permissionService.hasAllPermissions();
    if (hasAll && mounted) {
      // All permissions granted, navigate to onboarding
      Navigator.of(context).pushReplacementNamed('/onboarding');
      return;
    }

    final denied = await _permissionService.getDeniedPermissions();
    if (mounted) {
      setState(() {
        _deniedPermissions = denied;
      });
    }
  }

  Future<void> _requestPermissions() async {
    setState(() {
      _isRequesting = true;
    });

    try {
      final granted = await _permissionService.requestAllPermissions();

      if (!mounted) return;

      if (granted) {
        // All critical permissions granted, proceed to onboarding
        Navigator.of(context).pushReplacementNamed('/onboarding');
      } else {
        // Some permissions denied, show which ones
        final denied = await _permissionService.getDeniedPermissions();
        if (!mounted) return;
        
        setState(() {
          _deniedPermissions = denied;
        });

        // Show message about denied permissions
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
            content: Text(
              'Some permissions were denied. The app may not work properly without them.',
            ),
            backgroundColor: Colors.orange,
            duration: Duration(seconds: 4),
          ),
        );
      }
    } finally {
      if (mounted) {
        setState(() {
          _isRequesting = false;
        });
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(24.0),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: [
              const SizedBox(height: 40),
              Icon(
                Icons.security,
                size: 64,
                color: theme.colorScheme.primary,
              ),
              const SizedBox(height: 24),
              Text(
                'Permissions Required',
                style: theme.textTheme.headlineMedium?.copyWith(
                  fontWeight: FontWeight.bold,
                ),
                textAlign: TextAlign.center,
              ),
              const SizedBox(height: 12),
              Text(
                'Chord needs the following permissions to work properly:',
                style: theme.textTheme.bodyLarge?.copyWith(
                  color: Colors.grey[700],
                ),
                textAlign: TextAlign.center,
              ),
              const SizedBox(height: 24),
              _buildPermissionItem(
                icon: Icons.phone,
                title: 'Phone & Call Logs',
                description: 'To backup your call history to the cloud',
              ),
              const SizedBox(height: 12),
              _buildPermissionItem(
                icon: Icons.folder,
                title: 'Storage & Files',
                description: 'To access and backup your recordings',
              ),
              const SizedBox(height: 12),
              _buildPermissionItem(
                icon: Icons.notifications,
                title: 'Notifications',
                description: 'To notify you about sync status',
              ),
              const SizedBox(height: 12),
              _buildPermissionItem(
                icon: Icons.battery_charging_full,
                title: 'Battery Optimization',
                description: 'To keep background sync running',
              ),
              const SizedBox(height: 32),
              if (_deniedPermissions.isNotEmpty) ...[
                Container(
                  padding: const EdgeInsets.all(12),
                  decoration: BoxDecoration(
                    color: Colors.orange.shade50,
                    borderRadius: BorderRadius.circular(8),
                    border: Border.all(color: Colors.orange.shade200),
                  ),
                  child: Row(
                    children: [
                      Icon(Icons.warning, color: Colors.orange.shade700),
                      const SizedBox(width: 8),
                      Expanded(
                        child: Text(
                          'Missing: ${_deniedPermissions.join(", ")}',
                          style: TextStyle(
                            color: Colors.orange.shade700,
                            fontSize: 13,
                          ),
                        ),
                      ),
                    ],
                  ),
                ),
                const SizedBox(height: 12),
              ],
              ElevatedButton(
                onPressed: _isRequesting ? null : _requestPermissions,
                style: ElevatedButton.styleFrom(
                  backgroundColor: theme.colorScheme.primary,
                  foregroundColor: Colors.white,
                  padding: const EdgeInsets.symmetric(vertical: 16),
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(12),
                  ),
                ),
                child: _isRequesting
                    ? const SizedBox(
                        height: 20,
                        width: 20,
                        child: CircularProgressIndicator(
                          strokeWidth: 2,
                          valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                        ),
                      )
                    : const Text(
                        'Grant Permissions',
                        style: TextStyle(
                          fontSize: 16,
                          fontWeight: FontWeight.w600,
                        ),
                      ),
              ),
              const SizedBox(height: 24),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildPermissionItem({
    required IconData icon,
    required String title,
    required String description,
  }) {
    return Container(
      padding: const EdgeInsets.all(12),
      decoration: BoxDecoration(
        border: Border.all(color: Colors.grey.shade200),
        borderRadius: BorderRadius.circular(12),
      ),
      child: Row(
        children: [
          Container(
            padding: const EdgeInsets.all(10),
            decoration: BoxDecoration(
              color: Theme.of(context).colorScheme.primary.withAlpha(25),
              borderRadius: BorderRadius.circular(8),
            ),
            child: Icon(
              icon,
              color: Theme.of(context).colorScheme.primary,
              size: 22,
            ),
          ),
          const SizedBox(width: 12),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  title,
                  style: const TextStyle(
                    fontWeight: FontWeight.w600,
                    fontSize: 14,
                  ),
                ),
                const SizedBox(height: 2),
                Text(
                  description,
                  style: TextStyle(
                    color: Colors.grey[600],
                    fontSize: 12,
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

File: app/src/main/kotlin/com/even/chord/ui/screens/PermissionsScreen.kt
package com.even.chord.ui.screens

import android.Manifest
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.Settings
import android.widget.Toast
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material.icons.outlined.Warning
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import androidx.core.content.PermissionChecker
import com.even.chord.ui.theme.*

data class PermissionItem(
    val icon: ImageVector,
    val title: String,
    val description: String,
    val permissions: List<String>,
    val isSpecial: Boolean = false
)

@Composable
fun PermissionsScreen(
    onAllPermissionsGranted: () -> Unit
) {
    val context = LocalContext.current

    
    var isRequesting by remember { mutableStateOf(false) }
    var deniedPermissions by remember { mutableStateOf<List<String>>(emptyList()) }
    
    val permissionItems = remember {
        listOf(

            PermissionItem(
                icon = Icons.Default.Folder,
                title = "Storage & Files",
                description = "To access and backup your recordings",
                permissions = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                    listOf(Manifest.permission.MANAGE_EXTERNAL_STORAGE)
                } else {
                    listOf(
                        Manifest.permission.READ_EXTERNAL_STORAGE,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE
                    )
                },
                isSpecial = Build.VERSION.SDK_INT >= Build.VERSION_CODES.R
            ),
            PermissionItem(
                icon = Icons.Default.Notifications,
                title = "Notifications",
                description = "To notify you about sync status",
                permissions = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    listOf(Manifest.permission.POST_NOTIFICATIONS)
                } else {
                    emptyList()
                }
            ),
            PermissionItem(
                icon = Icons.Default.BatteryChargingFull,
                title = "Battery Optimization",
                description = "To keep background sync running",
                permissions = emptyList(),
                isSpecial = true
            )
        )
    }
    
    // Check permissions on load
    fun checkAllPermissions(): Boolean {

        val hasStorage = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            Environment.isExternalStorageManager()
        } else {
            ContextCompat.checkSelfPermission(
                context, Manifest.permission.READ_EXTERNAL_STORAGE
            ) == PermissionChecker.PERMISSION_GRANTED
        }
        
        // Return only storage status for now (though storage perms also removed in manifest, logic still here for legacy/scope?)
        // The user asked to drop READ_PHONE_STATE.
        return hasStorage
    }
    
    fun getDenied(): List<String> {
        val denied = mutableListOf<String>()
        

        val hasStorage = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            Environment.isExternalStorageManager()
        } else {
            ContextCompat.checkSelfPermission(
                context, Manifest.permission.READ_EXTERNAL_STORAGE
            ) == PermissionChecker.PERMISSION_GRANTED
        }
        if (!hasStorage) {
            denied.add("Storage & Files")
        }
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(
                    context, Manifest.permission.POST_NOTIFICATIONS
                ) != PermissionChecker.PERMISSION_GRANTED
            ) {
                denied.add("Notifications")
            }
        }
        
        return denied
    }
    
    LaunchedEffect(Unit) {
        if (checkAllPermissions()) {
            onAllPermissionsGranted()
        }
        deniedPermissions = getDenied()
    }
    
    val permissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { _ ->
        isRequesting = false
        deniedPermissions = getDenied()
        
        if (checkAllPermissions()) {
            onAllPermissionsGranted()
        } else {
            Toast.makeText(
                context,
                "Some permissions were denied. The app may not work properly.",
                Toast.LENGTH_LONG
            ).show()
        }
    }
    
    val manageStorageLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {
        deniedPermissions = getDenied()
        if (checkAllPermissions()) {
            onAllPermissionsGranted()
        }
    }
    
    Surface(
        modifier = Modifier.fillMaxSize(),
        color = Color.White
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(24.dp)
        ) {
            Spacer(modifier = Modifier.height(40.dp))
            
            Icon(
                imageVector = Icons.Default.Security,
                contentDescription = null,
                modifier = Modifier
                    .size(64.dp)
                    .align(Alignment.CenterHorizontally),
                tint = Primary
            )
            
            Spacer(modifier = Modifier.height(24.dp))
            
            Text(
                text = "Permissions Required",
                fontSize = 24.sp,
                fontWeight = FontWeight.Bold,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth()
            )
            
            Spacer(modifier = Modifier.height(12.dp))
            
            Text(
                text = "Chord needs the following permissions to work properly:",
                fontSize = 16.sp,
                color = Grey600,
                textAlign = TextAlign.Center,
                modifier = Modifier.fillMaxWidth()
            )
            
            Spacer(modifier = Modifier.height(24.dp))
            
            // Permission items
            permissionItems.forEach { item ->
                PermissionItemCard(item)
                Spacer(modifier = Modifier.height(12.dp))
            }
            
            // Denied permissions warning
            if (deniedPermissions.isNotEmpty()) {
                Spacer(modifier = Modifier.height(8.dp))
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(8.dp))
                        .background(Orange50)
                        .border(1.dp, Orange200, RoundedCornerShape(8.dp))
                        .padding(12.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Icon(
                        imageVector = Icons.Outlined.Warning,
                        contentDescription = null,
                        tint = Orange700,
                        modifier = Modifier.size(20.dp)
                    )
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = "Missing: ${deniedPermissions.joinToString(", ")}",
                        fontSize = 13.sp,
                        color = Orange700
                    )
                }
                Spacer(modifier = Modifier.height(12.dp))
            }
            
            Spacer(modifier = Modifier.height(20.dp))
            
            Button(
                onClick = {
                    isRequesting = true
                    
                    // Request MANAGE_EXTERNAL_STORAGE for Android 11+
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && 
                        !Environment.isExternalStorageManager()) {
                        val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
                        intent.data = Uri.parse("package:${context.packageName}")
                        manageStorageLauncher.launch(intent)
                    }
                    
                    // Phone permission removed
                    // val permissionsToRequest = mutableListOf<String>()
                    val permissionsToRequest = mutableListOf<String>()
                    // permissionsToRequest.add(Manifest.permission.READ_PHONE_STATE)
                    // permissionsToRequest.add(Manifest.permission.READ_CALL_LOG)
                    
                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.R) {
                        permissionsToRequest.add(Manifest.permission.READ_EXTERNAL_STORAGE)
                        permissionsToRequest.add(Manifest.permission.WRITE_EXTERNAL_STORAGE)
                    }
                    
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                        permissionsToRequest.add(Manifest.permission.POST_NOTIFICATIONS)
                    }
                    
                    permissionLauncher.launch(permissionsToRequest.toTypedArray())
                },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(52.dp),
                enabled = !isRequesting,
                colors = ButtonDefaults.buttonColors(
                    containerColor = Primary
                ),
                shape = RoundedCornerShape(12.dp)
            ) {
                if (isRequesting) {
                    CircularProgressIndicator(
                        modifier = Modifier.size(20.dp),
                        strokeWidth = 2.dp,
                        color = Color.White
                    )
                } else {
                    Text(
                        text = "Grant Permissions",
                        fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold
                    )
                }
            }
            
            Spacer(modifier = Modifier.height(24.dp))
        }
    }
}

@Composable
private fun PermissionItemCard(item: PermissionItem) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .border(1.dp, Grey200, RoundedCornerShape(12.dp))
            .padding(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(
            modifier = Modifier
                .size(42.dp)
                .clip(RoundedCornerShape(8.dp))
                .background(Primary.copy(alpha = 0.1f)),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = item.icon,
                contentDescription = null,
                tint = Primary,
                modifier = Modifier.size(22.dp)
            )
        }
        
        Spacer(modifier = Modifier.width(12.dp))
        
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = item.title,
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold
            )
            Spacer(modifier = Modifier.height(2.dp))
            Text(
                text = item.description,
                fontSize = 12.sp,
                color = Grey600
            )
        }
    }
}
